File: src/main/java/com/dersaun/apigestaocontas/infra/repository/LancamentoRepository.java
package com.dersaun.apigestaocontas.infra.repository;

import com.dersaun.apigestaocontas.domain.Lancamento;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.Optional;

public interface LancamentoRepository extends JpaRepository<Lancamento, Long> {

    @Query("""
        from Lancamento l 
        inner join fetch l.mes mes 
        inner join fetch l.pessoa pessoa 
        inner join fetch l.formaPagamento formaPagamento 
        left join fetch l.lancamentoRelacionado 
        left join fetch l.lancamentos lancamentos
        left join fetch lancamentos.mes mesLancamentos
        left join fetch lancamentos.pessoa pessoaLancamentos
        left join fetch lancamentos.formaPagamento formaPagamentoLancamentos 
        where l.id = :id
        
    """)
    Optional<Lancamento> buscarPorId(@Param("id") Long id);

    @Modifying()
    @Query("delete from Lancamento where lancamentoRelacionado = ?1")
    void deleteByLancamentoRelacionado(Lancamento lancamento);

}

File: src/app/conta/repositories/lancamento.repository.ts
import { ILancamentoRepository } from '@app/conta/interfaces/lancamento.repository.interface';
import { Lancamento } from '@app/conta/models/lancamento.entity';
import { Injectable } from '@nestjs/common';
import { EntityManager, Repository } from 'typeorm';

@Injectable()
export class LancamentoRepository
  extends Repository<Lancamento>
  implements ILancamentoRepository
{
  private cachedEntityManager: EntityManager;

  constructor(em: EntityManager) {
    super(Lancamento, em);
  }

  storeCurrentManager() {
    this.cachedEntityManager = this.manager;
  }
  restoreCurrentManager() {
    if (this.cachedEntityManager) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      this['manager'] = this.cachedEntityManager;
    }
  }
  setEntityManager(em: EntityManager): EntityManager {
    this.storeCurrentManager();
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    this['manager'] = em;
    return this.getEntityManager();
  }
  getEntityManager(): EntityManager {
    return this.manager;
  }

  async buscarCompleto(idLancamento: number): Promise<Lancamento> {
    const lancamento = await this.createQueryBuilder('lancamento')
      .leftJoinAndSelect('lancamento.mes', 'mes')
      .leftJoinAndSelect('lancamento.pessoa', 'pessoa')
      .leftJoinAndSelect('lancamento.formaPagamento', 'formaPagamento')
      .leftJoinAndSelect(
        'lancamento.lancamentoRelacionado',
        'lancamentoRelacionado',
      )
      .leftJoinAndSelect('lancamento.lancamentos', 'lancamentos')
      .leftJoinAndSelect('lancamentos.mes', 'mesLancamentos')
      .leftJoinAndSelect('lancamentos.pessoa', 'pessoaLancamentos')
      .leftJoinAndSelect(
        'lancamentos.formaPagamento',
        'formaPagamentoLancamentos',
      )
      .where('lancamento.id = :idLancamento', { idLancamento })
      .getOne();
    return lancamento;
  }

  async deleteByLancamentoRelacionado(lancamento: Lancamento): Promise<void> {
    const query = this.createQueryBuilder('lancamento')
      .leftJoin('lancamento.lancamentoRelacionado', 'lancamentoRelacionado')
      .delete()
      .where('lancamentoRelacionado.id = :idLancamento', {
        idLancamento: lancamento.id,
      });
    await query.execute();
  }
}
